### PR TITLE
Display appropriate import error messages for code size screen

### LINKED ISSUE #2215: When two identical snapshots are compared in the code size diff view, nothing happens
The following code results in a null value for `newRoot` which means the import diff UI stays on the screen with no indication that anything happened

```
final diffMap = buildComparisonTreemap(oldFile.data, newFile.data);
diffMap['n'] = 'Root';
final newRoot = generateDiffTree(diffMap);
```

### LINKED ISSUE #2244: Ensure we handle bad inputs to the code size screen
For example, we shouldn't accept non-json files or json files that don't match the format we expect (offline imports for the performance page for example)

### PR DESCRIPTION
**Pull request by peterdjlee**
Display appropriate import error messages for the code size screen
* SnapshotView
  * File type not supported.
    * When given a JSON that is not a snapshot or an APK analysis.
* DiffView
  * File type not supported.
    * When given a JSON that is not a snapshot or an APK analysis.
  * Given snapshots to compare are of different types.
    * When given two JSONs of different types (e.g. APK analysis vs snapshot).
  * Given snapshots to compare are identical. 
    * When given two JSONs that have the same data.

Fixes #2244 
Fixes #2215

### CODE REVIEW

File: packages/devtools_app/lib/src/code_size/code_size_controller.dart [hunk 1]
```diff
@@ -118,16 +119,25 @@ class CodeSizeController {
...
     Map<String, dynamic> processedJson;
     if (jsonFile.isApkFile) {
       // APK analysis json should be processed already.
       processedJson = jsonFile.data;
     } else {
-      processedJson = treemapFromJson(jsonFile.data);
+      try {
+        processedJson = treemapFromJson(jsonFile.data);
+      } catch (error) {
+        onError('Failed to load snapshot: File type not supported.');
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/code_size/code_size_controller.dart:135]: let's add some more details:

"Failed to load snapshot: file type not supported.

The code size tool supports Dart AOT v8 snapshots, instruction sizes, and apk-analysis files. See documentation for how to generate these files."

File: packages/devtools_app/lib/src/code_size/code_size_controller.dart [hunk 2]
```diff
@@ -141,11 +151,17 @@ class CodeSizeController {
...
-    if (oldFile == null || newFile == null) return;
+    if (oldFile == null || newFile == null) {
+      return;
+    }
 
-    changeOldDiffSnapshotFile(oldFile);
-    changeNewDiffSnapshotFile(newFile);
+    if (oldFile.isApkFile != newFile.isApkFile) {
+      onError(
+        'Failed to load snapshot: Given snapshots to compare are of different types.',
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/code_size/code_size_controller.dart:162]: "Failed to load diff: OLD and NEW files are different types."

File: packages/devtools_app/lib/src/code_size/code_size_controller.dart [hunk 3]
```diff
@@ -141,11 +151,17 @@ class CodeSizeController {
...
+    void Function(String error) onError,
   ) {
-    if (oldFile == null || newFile == null) return;
+    if (oldFile == null || newFile == null) {
+      return;
+    }
 
-    changeOldDiffSnapshotFile(oldFile);
-    changeNewDiffSnapshotFile(newFile);
+    if (oldFile.isApkFile != newFile.isApkFile) {
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/code_size/code_size_controller.dart:160]: what about v8 snapshot vs instruction sizes? we should probably add a check for that too

File: packages/devtools_app/lib/src/code_size/code_size_controller.dart [hunk 4]
```diff
@@ -165,9 +181,24 @@ class CodeSizeController {
 
       diffMap = compareProgramInfo(oldApkProgramInfo, newApkProgramInfo);
     } else {
-      diffMap = buildComparisonTreemap(oldFile.data, newFile.data);
+      try {
+        diffMap = buildComparisonTreemap(oldFile.data, newFile.data);
+      } catch (error) {
+        onError('Failed to load snapshot: File type not supported.');
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/code_size/code_size_controller.dart:187]: "Failed to load diff: file type not supported. 

The code size tool supports AOT v8 snapshots, instruction sizes, and apk-analysis files. See documentation for how to generate these files."

File: packages/devtools_app/lib/src/code_size/code_size_controller.dart [hunk 5]
```diff
@@ -165,9 +181,24 @@ class CodeSizeController {
...
+        return;
+      }
+    }
+
+    if ((diffMap['children'] as List).isEmpty) {
+      onError(
+        'Failed to load snapshot: Given snapshots to compare are identical.',
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/code_size/code_size_controller.dart:194]: "Failed to load diff: OLD and NEW files are identical."

File: packages/devtools_app/lib/src/code_size/file_import_container.dart [hunk 6]
```diff
@@ -263,7 +268,12 @@ class _DualFileImportContainerState extends State<DualFileImportContainer> {
           children: [
             RaisedButton(
               onPressed: firstImportedFile != null && secondImportedFile != null
-                  ? () => widget.onAction(firstImportedFile, secondImportedFile)
+                  ? () => widget.onAction(
+                        firstImportedFile,
+                        secondImportedFile,
+                        (errorMessage) =>
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/code_size/file_import_container.dart:274]: rename to error to be consistent with snapshot callback

File: packages/devtools_app/lib/src/code_size/code_size_controller.dart (hunk 1 above)
Comment (peterdjlee) [reply packages/devtools_app/lib/src/code_size/code_size_controller.dart]: Done

File: packages/devtools_app/lib/src/code_size/code_size_controller.dart (hunk 3 above)
Comment (peterdjlee) [reply packages/devtools_app/lib/src/code_size/code_size_controller.dart]: Done

File: packages/devtools_app/lib/src/code_size/code_size_controller.dart (hunk 2 above)
Comment (peterdjlee) [reply packages/devtools_app/lib/src/code_size/code_size_controller.dart]: Done

File: packages/devtools_app/lib/src/code_size/code_size_controller.dart (hunk 4 above)
Comment (peterdjlee) [reply packages/devtools_app/lib/src/code_size/code_size_controller.dart]: Done

File: packages/devtools_app/lib/src/code_size/code_size_controller.dart (hunk 5 above)
Comment (peterdjlee) [reply packages/devtools_app/lib/src/code_size/code_size_controller.dart]: Done

File: packages/devtools_app/lib/src/code_size/file_import_container.dart (hunk 6 above)
Comment (peterdjlee) [reply packages/devtools_app/lib/src/code_size/file_import_container.dart]: Done

File: packages/devtools_app/lib/src/code_size/code_size_controller.dart [hunk 7]
```diff
@@ -118,17 +119,27 @@ class CodeSizeController {
...
     Map<String, dynamic> processedJson;
     if (jsonFile.isApkFile) {
       // APK analysis json should be processed already.
       processedJson = jsonFile.data;
     } else {
-      processedJson = treemapFromJson(jsonFile.data);
+      try {
+        processedJson = treemapFromJson(jsonFile.data);
+      } catch (error) {
+        onError('Failed to load snapshot: file type not supported.\n\n'
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/code_size/code_size_controller.dart:134]: add todo to link to docs, here and in the diff version of this message

File: packages/devtools_app/lib/src/code_size/code_size_controller.dart [hunk 8]
```diff
@@ -165,9 +184,26 @@ class CodeSizeController {
 
       diffMap = compareProgramInfo(oldApkProgramInfo, newApkProgramInfo);
     } else {
-      diffMap = buildComparisonTreemap(oldFile.data, newFile.data);
+      try {
+        diffMap = buildComparisonTreemap(oldFile.data, newFile.data);
+      } catch (error) {
+        onError('Failed to load snapshot: file type not supported.\n\n'
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/code_size/code_size_controller.dart:190]: r/snapshot/diff

File: packages/devtools_app/lib/src/code_size/code_size_controller.dart [hunk 9]
```diff
@@ -165,9 +184,26 @@ class CodeSizeController {
...
+            'The code size tool supports Dart AOT v8 snapshots, instruction sizes, '
+            'and apk-analysis files. See documentation for how to generate these files.');
+        return;
+      }
     }
 
+    if ((diffMap['children'] as List).isEmpty) {
+      onError(
+        'Failed to load diff: OLD and NEW files are identical.',
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/code_size/code_size_controller.dart:199]: remove trailing comma here - it should fit on one line

File: packages/devtools_app/lib/src/code_size/code_size_controller.dart [hunk 10]
```diff
@@ -339,6 +375,14 @@ extension CodeSizeJsonFileExtension on DevToolsJsonFile {
     return false;
   }
 
+  bool get isV8Snapshot {
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/code_size/code_size_controller.dart:378]: can we use the isV8Snapshot utility from package:vm_snapshot_analysis

Comment (kenzieschmoll) [reply packages/devtools_app/lib/src/code_size/code_size_controller.dart:380]: that's weird. Have you debugged this locally to see what's happening?

Comment (peterdjlee) [reply packages/devtools_app/lib/src/code_size/code_size_controller.dart]: Looks like I was in fact not passing in the expected data. It's working now.

**Comment by peterdjlee** [comment]
Would it be good to add tests to verify the error notification shows given invalid files?

**Comment by kenzieschmoll** [comment]
yes we should add tests

File: packages/devtools_app/lib/src/code_size/code_size_screen.dart [hunk 11]
```diff
@@ -274,34 +275,38 @@ class _SnapshotViewState extends State<SnapshotView> with AutoDisposeMixin {
...
+              children: [
+                Flexible(
+                  child: FileImportContainer(
+                    title: 'Snapshot / APK analysis',
+                    instructions: SnapshotView.importInstructions,
+                    actionText: 'Analyze Snapshot / APK',
+                    onAction: (jsonFile) {
+                      controller.loadTreeFromJsonFile(
+                        jsonFile,
+                        (error) => Notifications.of(context).push(error),
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/code_size/code_size_screen.dart:301]: we need to store Notifications.of(context) in a var before the callback here too

File: packages/devtools_app/test/support/code_size_test_data/unsupported_file.dart [hunk 12]
```diff
@@ -0,0 +1,5 @@
+const unsupportedFile = '''
+{
+  "description": "JSON to test error output when importing an unsupported JSON."
+}
+''';
```
Comment (kenzieschmoll) [inline packages/devtools_app/test/support/code_size_test_data/unsupported_file.dart:5]: new line at eof

File: packages/devtools_app/lib/src/code_size/code_size_screen.dart (hunk 11 above)
Comment (peterdjlee) [reply packages/devtools_app/lib/src/code_size/code_size_screen.dart:301]: Done

File: packages/devtools_app/test/support/code_size_test_data/unsupported_file.dart (hunk 12 above)
Comment (peterdjlee) [reply packages/devtools_app/test/support/code_size_test_data/unsupported_file.dart:5]: Done

**Pull request merged**